src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncStateHandler.scala

```diff
@@ -0,0 +1,446 @@
+        _ <- blockchain.getBlockBodyByHash(hash)
+        _ <- blockchain.getReceiptsByHash(hash)
+      } yield header
+    }
+
+    if (fullBlocks.nonEmpty) {
+      val bestReceivedBlock = fullBlocks.maxBy(_.number)
+      if (appStateStorage.getBestBlockNumber() < bestReceivedBlock.number) {
+        appStateStorage.putBestBlockNumber(bestReceivedBlock.number)
+      }
+    }
+
+  }
+
+  private def validHeaders(syncState: SyncState, blockHeaders: Seq[BlockHeader], requestedNum: BigInt): Boolean =
+    blockHeaders.nonEmpty &&
+      blockHeaders.size <= requestedNum &&
+      blockHeaders.head.number == (syncState.bestBlockHeaderNumber + 1) &&
+      checkHeadersChain(blockHeaders)
+
+  private def updateTargetSyncState(syncState: SyncState, state: FinalBlockProcessingResult, targetBlockHeader: BlockHeader): SyncState = state match {
+    case ImportedLastBlock =>
+      if (targetBlockHeader.number - syncState.targetBlock.number <= syncConfig.maxTargetDifference) {
+        log.info(s"Current target block is fresh enough, starting state download")
+        syncState.copy(pendingMptNodes = Seq(StateMptNodeHash(syncState.targetBlock.stateRoot)))
+      } else {
+        log.info(s"Changing target block to ${targetBlockHeader.number}, new safe target is ${targetBlockHeader.number + syncConfig.fastSyncBlockValidationX}")
+        syncState.updateTargetBlock(targetBlockHeader, syncConfig.fastSyncBlockValidationX, updateFailures = false)
+      }
+
+    case LastBlockValidationFailed(_) =>
+      log.info(s"Changing target block after failure, to ${targetBlockHeader.number}, new safe target is ${syncState.safeDownloadTarget}")
+      syncState.updateTargetBlock(targetBlockHeader, syncConfig.fastSyncBlockValidationX, updateFailures = true)
+  }
+
+  private def handleBlockHeaders(syncState: SyncState, peer: Peer, headers: Seq[BlockHeader]): (FastSyncCommand, SyncState) = {
+    processHeaders(syncState, peer, headers) match {
+      case ParentDifficultyNotFound(header, finalState) =>
+        log.debug("Parent difficulty not found for block {}, not processing rest of headers", header.number)
+        (ContinueSyncing(None), finalState)
+      case HeadersProcessingFinished(finalState) =>
+        (ContinueSyncing(None), finalState)
+      case ImportedTargetBlock(finalState)  =>
+        (InitTargetBlockUpdate(ImportedLastBlock), finalState)
+      case ValidationFailed(header, peerToBlackList, finalState) =>
+        (handleBlockValidationError(finalState, header, peerToBlackList), finalState)
+    }
+  }
+
+  @tailrec
+  private def processHeaders(syncState: SyncState, peer: Peer, headers: Seq[BlockHeader]): HeaderProcessingResult = {
+    if (headers.nonEmpty) {
+      val header = headers.head
+      processHeader(syncState, header, peer) match {
+        case Left(result)        => result
+        case Right(headerAndDif) =>
+          val newState = updateSyncState(headerAndDif._1, headerAndDif._2)
```

Comment:
use pattern matching instead calling ._1 and _.2

Reply:
hmm i am not convinced that adding pattern matching will increase readability here.
there are only two params in tuple, and also called function is really near.